manager/assets/modext/widgets/core/modx.grid.js

```diff
@@ -495,14 +515,190 @@ Ext.extend(MODx.grid.Grid,Ext.grid.EditorGridPanel,{
+        return !Ext.isEmpty(permissions) && !record.json.isProtected && permissions.delete === true ? true : false ;
+    }
+
+    ,userCanDuplicateRecord: function(record) {
+        const permissions = record.json.permissions;
+        return !Ext.isEmpty(permissions) && permissions.duplicate === true ? true : false ;
+    }
+
+    /**
+     * @property {Function} setShowActionsMenu - Based on properties set in the calling child class and the
+     * the current user's permissions for actions taken within that class (create, edit, delete, etc),
+     * evaluates whether the actions menu trigger should appear and sets boolean value on the showActionsMenu property
+     *
+     * @return void
+     */
+    ,setShowActionsMenu: function() {
+        if (this.config.disableContextMenuAction === true) {
+            this.showActionsMenu = false;
+            return;
+        }
+        const permissions = [];
+        this.gridMenuActions.forEach(mode => {
+            mode = mode == 'duplicate' ? 'userCanCreate' : 'userCan' + Ext.util.Format.capitalize(mode);
+            const modePermission = mode === 'userCanExport' ? true : this[mode] ;
+            if (['userCanCreate', 'userCanEdit'].includes(mode) && modePermission === true) {
+                this.hasSavePermissions = true;
+            }
+            permissions.push(modePermission);
+        });
+        this.showActionsMenu = permissions.length === 0 || permissions.every(permission => permission === false) === true
+            ? false
+            : true
+            ;
+    }
+
+    /**
+     * @property {Function} recordIsProtected - Used to remove the ability to delete
+     * specific record rows, regardless of permissions levels, based on a given record identifier
+     *
+     * @param {Number} subject - The value of the current record's identifier
+     * @param {Number} protectedIdentifiers - The record identifiers to be protected (making them non-editable/deletable)
+     *
+     * @return {Boolean}
+     */
+    ,recordIsProtected: function(subject, protectedIdentifiers) {
+        if (Ext.isEmpty(protectedIdentifiers)) {
+            return false;
+        }
+        protectedIdentifiers = protectedIdentifiers.map(identifier => {
+            return typeof identifier === 'string' ? identifier.trim() : identifier ;
+        });
+        return protectedIdentifiers.includes(subject);
+    }
+
+    /**
+     * @property {Function} valueIsReserved - Wraps a grid value with a real or simulated link — a trigger that appears
+     * like an anchor link, usually to access a dropdown chooser or other control
+     *
+     * @param {Array|String} reservedValues - A set of values that can not be used for a particular object's field
+     * @param {Object} value - The submitted value being tested
+     *
+     * @return {Boolean}
+     */
+    ,valueIsReserved: function(reservedValues, value) {
+        if (!Ext.isArray(reservedValues)) {
```

Comment:
```suggestion
        if (!Array.isArray(reservedValues)) {
```

Reply:
<img width="766" alt="Снимок экрана 2021-11-25 в 19 44 12" src="https://user-images.githubusercontent.com/5341208/143478311-4e6c00ce-fe3e-4ed9-b449-13108010d1a4.png">

Reply:
I had been deferring to Ext's polyfills, but you're probably right that this one's not needed since the native isArray is widely compatible and the Ext version does nothing special (looking at both 3.4 and 7.x).